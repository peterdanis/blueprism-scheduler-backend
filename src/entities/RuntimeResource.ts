import { Column, Entity } from "typeorm";
import Base from "./Base";

@Entity()
export default class RuntimeResource extends Base {
  @Column()
  hostname!: string;

  @Column()
  friendlyName!: string;
}
