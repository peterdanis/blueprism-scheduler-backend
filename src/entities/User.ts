import { Column, Entity } from "typeorm";
import Base from "./Base";

@Entity()
export default class User extends Base {
  @Column({ unique: true })
  name!: string;

  @Column({ default: true })
  readonly!: boolean;
}
