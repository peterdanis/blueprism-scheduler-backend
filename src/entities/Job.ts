import { Entity, ManyToOne } from "typeorm";
import JobBase from "./JobBase";
import RuntimeResource from "./RuntimeResource";
import Schedule from "./Schedule";

@Entity()
export default class Job extends JobBase {
  @ManyToOne(() => RuntimeResource, (runtimeResource) => runtimeResource.job, {
    eager: true,
    nullable: false,
  })
  runtimeResource!: RuntimeResource;

  @ManyToOne(() => Schedule, (schedule) => schedule.job, {
    eager: true,
    nullable: false,
  })
  schedule!: Schedule;
}
