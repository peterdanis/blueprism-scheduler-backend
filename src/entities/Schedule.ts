import { Column, Entity, ManyToOne, OneToMany } from "typeorm";
import Base from "./Base";
import { defaultPriority } from "../utils/getSetting";
import Job from "./Job";
import RuntimeResource from "./RuntimeResource";
import ScheduleTask from "./ScheduleTask";

interface OnError {
  action: "email";
  emailTo?: string;
  emailCc?: string;
}

@Entity()
export default class Schedule extends Base {
  @Column({ default: false })
  force!: boolean;

  @Column({ unique: true })
  name!: string;

  @Column({ default: 1800000 }) // The default 1800000ms equals to 30min
  hardForceTime!: number;

  @Column({ default: 86400000 }) // The default 86400000ms equals to 1d
  hardTimeout!: number;

  @Column({ nullable: true, type: "simple-json" })
  onError?: OnError;

  @Column({ default: defaultPriority })
  priority!: number;

  @Column()
  rule!: string;

  @Column({ default: 900000 }) // The default 900000ms equals to 15min
  softForceTime!: number;

  @Column({ default: 86400000 }) // The default 86400000ms equals to 1d
  softTimeout!: number;

  @Column()
  validFrom!: Date;

  @Column({ default: new Date(253402214400000).toISOString() }) // Equals to 31st of December 9999
  validUntil!: Date;

  @Column({ default: 86400000 }) // The default 86400000ms equals to 1d
  waitTime!: number;

  @OneToMany(() => ScheduleTask, (scheduleTask) => scheduleTask.schedule, {
    cascade: true,
    eager: true,
  })
  scheduleTask!: ScheduleTask[];

  @OneToMany(() => Job, (job) => job.schedule)
  job!: Job[];

  @ManyToOne(
    () => RuntimeResource,
    (runtimeResource) => runtimeResource.schedule,
    {
      eager: true,
      nullable: false,
    },
  )
  runtimeResource!: RuntimeResource;
}
