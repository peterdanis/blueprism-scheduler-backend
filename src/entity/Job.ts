import { Column, Entity, ManyToOne } from "typeorm";
import Base from "./Base";
import { defaultPriority } from "../utils/getSetting";
import RuntimeResource from "./RuntimeResource";
import Schedule from "./Schedule";

type Status = "waiting" | "running" | "finished" | "failed";

@Entity()
export default class Job extends Base {
  @Column({ default: defaultPriority })
  priority!: number; // De-normalized priority field - to by able to get the queued schedule higher or lower priority

  @Column({ nullable: true })
  sessionId!: string;

  @Column()
  startTime!: Date;

  @Column()
  status!: Status;

  @Column({ default: 0 })
  step!: number;

  @Column({ default: 0 })
  subStep!: number;

  @ManyToOne(() => RuntimeResource, (runtimeResource) => runtimeResource.job, {
    nullable: false,
  })
  runtimeResource!: RuntimeResource;

  @ManyToOne(() => Schedule, (schedule) => schedule.job, {
    nullable: false,
  })
  schedule!: Schedule;
}
