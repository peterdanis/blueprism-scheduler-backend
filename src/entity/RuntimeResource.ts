import { Column, Entity, OneToMany } from "typeorm";
import Base from "./Base";
import Job from "./Job";
import Schedule from "./Schedule";

type Auth = "basic" | "apikey";

@Entity()
export default class RuntimeResource extends Base {
  @Column({ nullable: true })
  apiKey?: string;

  @Column()
  auth!: Auth;

  @Column()
  friendlyName!: string;

  @Column()
  hostname!: string;

  @Column({ default: true })
  https!: boolean;

  @Column({ nullable: true })
  password?: string;

  @Column()
  port!: number;

  @Column({ nullable: true })
  username?: string;

  @OneToMany(() => Job, (job) => job.schedule)
  job!: Job[];

  @OneToMany(() => Schedule, (schedule) => schedule.runtimeResource)
  schedule!: Schedule[];
}
