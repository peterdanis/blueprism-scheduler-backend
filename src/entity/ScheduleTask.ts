import { Column, Entity, ManyToOne, Unique } from "typeorm";
import Base from "./Base";
import Schedule from "./Schedule";
import Task from "./Task";

@Entity()
@Unique("step_scheduleId", ["step", "schedule"])
export default class ScheduleTask extends Base {
  @Column({ default: 0 })
  delayAfter!: number;

  @Column()
  step!: number;

  @ManyToOne(() => Schedule, (schedule) => schedule.scheduleTask, {
    nullable: false,
  })
  schedule!: Schedule;

  @ManyToOne(() => Task, (task) => task.scheduleTask, {
    eager: true,
    nullable: false,
  })
  task!: Task;
}
