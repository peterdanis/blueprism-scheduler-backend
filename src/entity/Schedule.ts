import { Column, Entity, ManyToOne, OneToMany } from "typeorm";
import Base from "./Base";
import { defaultPriority } from "../utils/getSetting";
import Job from "./Job";
import RuntimeResource from "./RuntimeResource";
import ScheduleTask from "./ScheduleTask";

@Entity()
export default class Schedule extends Base {
  @Column({ unique: true })
  name!: string;

  @Column({ default: defaultPriority })
  priority!: number;

  @Column()
  rule!: string;

  @Column()
  validFrom!: Date;

  @Column({ default: new Date(253402214400000).toISOString() }) // Equals to 31st of December 9999
  validUntil!: Date;

  @OneToMany(() => ScheduleTask, (scheduleTask) => scheduleTask.schedule)
  scheduleTask!: ScheduleTask[];

  @OneToMany(() => Job, (job) => job.schedule)
  job!: Job[];

  @ManyToOne(
    () => RuntimeResource,
    (runtimeResource) => runtimeResource.schedule,
    {
      eager: true,
      nullable: false,
    },
  )
  runtimeResource!: RuntimeResource;
}
