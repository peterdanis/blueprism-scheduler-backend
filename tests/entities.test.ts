import { Connection, createConnection } from "typeorm";
import config from "./testUtils/testConnectionConfig";
import Job from "../src/entity/Job";
import loadDummies from "./testUtils/loadDummies";
import RuntimeResource from "../src/entity/RuntimeResource";
import Schedule from "../src/entity/Schedule";
import ScheduleTask from "../src/entity/ScheduleTask";
import Task from "../src/entity/Task";
import User from "../src/entity/User";

let connection: Connection;
let runtimeResource: RuntimeResource | undefined;
let schedule: Schedule | undefined;

beforeAll(async () => {
  connection = await createConnection(config);
  await loadDummies();
});

afterAll(async () => {
  await connection.close();
});

describe("User", () => {
  test("has expected properties", async () => {
    const users = await User.find();
    expect(users[0]).toMatchSnapshot();
  });
});

describe("RuntimeResource", () => {
  test("has expected properties", async () => {
    const runtimeResources = await RuntimeResource.find();
    [runtimeResource] = runtimeResources;
    expect(runtimeResource).toMatchSnapshot();
  });
});

describe("Schedule", () => {
  test("has expected properties", async () => {
    const schedules = await Schedule.find();
    [schedule] = schedules;
    expect(schedule).toMatchSnapshot();
  });
});

describe("Task", () => {
  test("has expected properties", async () => {
    const tasks = await Task.find();
    expect(tasks[0]).toMatchSnapshot();
  });
});

describe("ScheduleTask", () => {
  test("has expected properties", async () => {
    const scheduleTasks = await ScheduleTask.find();
    expect(scheduleTasks[0]).toMatchSnapshot();
  });
});

describe("Job", () => {
  test("has expected properties", async () => {
    const job = Job.create({
      addTime: new Date("2020-12-31T20:00:00Z"),
      runtimeResource,
      schedule,
      startReason: "test",
      status: "waiting",
      updateTime: new Date("2020-12-31T20:00:00Z"),
    });
    await job.save();
    const jobs = await Job.find();
    expect(jobs[0]).toMatchSnapshot();
  });
});
