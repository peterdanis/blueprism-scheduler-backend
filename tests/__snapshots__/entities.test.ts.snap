// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Job has expected properties 1`] = `
Job {
  "addTime": 2020-12-31T20:00:00.000Z,
  "endTime": null,
  "id": 1,
  "message": null,
  "priority": 50,
  "runtimeResource": RuntimeResource {
    "apiKey": null,
    "auth": "basic",
    "friendlyName": "vm1",
    "hostname": "vm1.something.com",
    "https": true,
    "id": 1,
    "password": "p@ssword#1",
    "port": 3000,
    "username": "rpa",
  },
  "schedule": Schedule {
    "force": false,
    "hardForceTime": 1800000,
    "hardTimeout": 86400000,
    "id": 1,
    "name": "Test schedule 1",
    "onError": null,
    "priority": 50,
    "rule": "* * * * *",
    "runtimeResource": RuntimeResource {
      "apiKey": null,
      "auth": "basic",
      "friendlyName": "vm1",
      "hostname": "vm1.something.com",
      "https": true,
      "id": 1,
      "password": "p@ssword#1",
      "port": 3000,
      "username": "rpa",
    },
    "scheduleTask": Array [
      ScheduleTask {
        "abortEarly": true,
        "delayAfter": 0,
        "id": 4,
        "onError": null,
        "step": 2,
        "task": Task {
          "hardTimeout": 150000,
          "id": 2,
          "inputs": Array [
            Object {
              "@name": "Time to run",
              "@type": "text",
              "@value": "120",
            },
          ],
          "name": "Test process 2mins",
          "process": "Test process",
          "softTimeout": 150000,
        },
      },
      ScheduleTask {
        "abortEarly": true,
        "delayAfter": 0,
        "id": 7,
        "onError": null,
        "step": 3,
        "task": Task {
          "hardTimeout": 350000,
          "id": 3,
          "inputs": Array [
            Object {
              "@name": "Time to run",
              "@type": "text",
              "@value": "300",
            },
          ],
          "name": "Test process 5mins",
          "process": "Test process",
          "softTimeout": 350000,
        },
      },
      ScheduleTask {
        "abortEarly": true,
        "delayAfter": 30000,
        "id": 1,
        "onError": null,
        "step": 1,
        "task": Task {
          "hardTimeout": 100000,
          "id": 1,
          "inputs": null,
          "name": "Login",
          "process": "Login",
          "softTimeout": 100000,
        },
      },
    ],
    "softForceTime": 900000,
    "softTimeout": 86400000,
    "validFrom": 2020-12-31T20:00:00.000Z,
    "validUntil": 9999-12-31T00:00:00.000Z,
    "waitTime": 86400000,
  },
  "sessionId": null,
  "startReason": "test",
  "startTime": null,
  "status": "waiting",
  "step": 1,
  "steps": null,
  "stopReason": null,
  "subStep": 1,
  "updateTime": 2020-12-31T20:00:00.000Z,
}
`;

exports[`RuntimeResource has expected properties 1`] = `
RuntimeResource {
  "apiKey": null,
  "auth": "basic",
  "friendlyName": "vm1",
  "hostname": "vm1.something.com",
  "https": true,
  "id": 1,
  "password": "p@ssword#1",
  "port": 3000,
  "username": "rpa",
}
`;

exports[`Schedule has expected properties 1`] = `
Schedule {
  "force": false,
  "hardForceTime": 1800000,
  "hardTimeout": 86400000,
  "id": 1,
  "name": "Test schedule 1",
  "onError": null,
  "priority": 50,
  "rule": "* * * * *",
  "runtimeResource": RuntimeResource {
    "apiKey": null,
    "auth": "basic",
    "friendlyName": "vm1",
    "hostname": "vm1.something.com",
    "https": true,
    "id": 1,
    "password": "p@ssword#1",
    "port": 3000,
    "username": "rpa",
  },
  "scheduleTask": Array [
    ScheduleTask {
      "abortEarly": true,
      "delayAfter": 0,
      "id": 4,
      "onError": null,
      "step": 2,
      "task": Task {
        "hardTimeout": 150000,
        "id": 2,
        "inputs": Array [
          Object {
            "@name": "Time to run",
            "@type": "text",
            "@value": "120",
          },
        ],
        "name": "Test process 2mins",
        "process": "Test process",
        "softTimeout": 150000,
      },
    },
    ScheduleTask {
      "abortEarly": true,
      "delayAfter": 0,
      "id": 7,
      "onError": null,
      "step": 3,
      "task": Task {
        "hardTimeout": 350000,
        "id": 3,
        "inputs": Array [
          Object {
            "@name": "Time to run",
            "@type": "text",
            "@value": "300",
          },
        ],
        "name": "Test process 5mins",
        "process": "Test process",
        "softTimeout": 350000,
      },
    },
    ScheduleTask {
      "abortEarly": true,
      "delayAfter": 30000,
      "id": 1,
      "onError": null,
      "step": 1,
      "task": Task {
        "hardTimeout": 100000,
        "id": 1,
        "inputs": null,
        "name": "Login",
        "process": "Login",
        "softTimeout": 100000,
      },
    },
  ],
  "softForceTime": 900000,
  "softTimeout": 86400000,
  "validFrom": 2020-12-31T20:00:00.000Z,
  "validUntil": 9999-12-31T00:00:00.000Z,
  "waitTime": 86400000,
}
`;

exports[`ScheduleTask has expected properties 1`] = `
ScheduleTask {
  "abortEarly": true,
  "delayAfter": 30000,
  "id": 1,
  "onError": null,
  "step": 1,
  "task": Task {
    "hardTimeout": 100000,
    "id": 1,
    "inputs": null,
    "name": "Login",
    "process": "Login",
    "softTimeout": 100000,
  },
}
`;

exports[`Task has expected properties 1`] = `
Task {
  "hardTimeout": 100000,
  "id": 1,
  "inputs": null,
  "name": "Login",
  "process": "Login",
  "softTimeout": 100000,
}
`;

exports[`User has expected properties 1`] = `
User {
  "admin": false,
  "id": 1,
  "name": "testuser",
}
`;
